src/backend/renderer/sync/mod.rs

```diff
@@ -0,0 +1,90 @@
+//! Helper for synchronizing rendering operations
+use std::{os::unix::io::OwnedFd, sync::Arc};
+
+use downcast_rs::{impl_downcast, Downcast};
+
+#[cfg(feature = "backend_egl")]
+mod egl;
+
+/// A fence that will be signaled in finite time
+pub trait Fence: std::fmt::Debug + Send + Sync + Downcast {
+    /// Queries the state of the fence
+    fn is_signaled(&self) -> bool;
+
+    /// Blocks the current thread until the fence is signaled.
+    fn wait(&self);
+
+    /// Returns whether this fence has a native handle
+    fn is_native(&self) -> bool;
+
+    /// Get the native fence fd if available
+    fn native(&self) -> Option<OwnedFd>;
```

Comment:
What are the expected semantics this function should consider?

I see a few things here:
1. Every call to `Fence::native` must return a new fd. This seems to imply that implementations should export the sync fd when this function is called.
2. One possible implementation strategy I could think of is to export the fd before returning the sync point and just dup the fd. However `dup`ing the fd returns an io::Result.

Another question: could we return a `BorrowedFd` here and be fine? I am not sure if the fd needs to be valid for the lifetime of the `IN_FENCES_PTR` or whatever it is called or could you just drop the fence immediately after commiting/page flip?

Reply:
Yes, 1. is correct, currently it expects to always return a new fd as indicated by returning an `OwnedFd` (of course this is some kind of implementation detail as the `Fence` could also dup the fd). For the `Fence` trait this is correct imo as both, EGLFence and VkFence, export a new fd and transfer ownership.
Personally I don't think 2. would be a good fit, we do not need a native fd for all use-cases. If you just want to insert the fence in another renderer (as a server wait) or blocking wait we do not necessarily need the fd at all. In case we want to import a fd into a renderer we again have to transfer ownership and I personally don't want to introduce another error case by `dup` here.

`IN_FENCE_FD` already works with `BorrowedFd` in the drm abstraction. In case of linux the native fence (the fd) is expected to point to a `sync_file`. On a test commit kms only checks if the provided fd points to a valid `sync_file`, on commit it will import the sync_file internally. But I opted to cache the `OwnedFd` in the `DrmCompositor` so that we do not have to re-export the fd on each atomic test.